pkg/serial/handshake.go

```diff
@@ -28,134 +47,97 @@ import (
 )
 
 const (
-	flagSyn      byte = 0x16
-	flagAck      byte = 0x06
-	flagDebugAck byte = 0x07
-	flagNak      byte = 0x15
+	flagSyn byte = 0x16
+	flagAck      = 0x06
+	flagNak      = 0x15
 )
 
 // HandshakeError should only occure if the protocol between HandshakeServer and HandshakeClient was violated.
 type HandshakeError struct {
 	msg string
 }
 
-func (he *HandshakeError) Error() string { return "HandshakeServer: " + he.msg }
+func (he *HandshakeError) Error() string {
+	return he.msg
+}
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
 
-// PurgeIncoming is used to clear a channel of bytes prior to handshaking
-func PurgeIncoming(conn net.Conn) {
-	if tracing {
-		defer trace.End(trace.Begin(""))
+// ReadAtLeastN reads at least l bytes and returns those l bytes or errors
+func ReadAtLeastN(conn io.ReadWriter, buffer []byte, l int) ([]byte, error) {
+	n, err := io.ReadAtLeast(conn, buffer, l)
+	if err != nil {
+		return nil, err
 	}
-	buf := make([]byte, 255)
 
-	// read until the incoming channel is empty
-	conn.SetReadDeadline(time.Now().Add(time.Duration(10 * time.Millisecond)))
-	for n, err := conn.Read(buf); n != 0 || err == nil; n, err = conn.Read(buf) {
-		log.Debugf("discarding following %d bytes from input channel\n", n)
-		log.Debugf("%+v\n", buf[0:n])
+	// however if we read more than l, it means buffer is not empty
+	if n != l {
+		buffer = buffer[n-l:]
```

Comment:
I think it doesn't matter whether it is real garbage or not. Client and server will get in sync eventually even if we drop legitimate handshake packages as garbage or read in the middle of a write etc. This is only for our handshake so after that SSH will take over and will do its own thing so there is nothing valuable to lose in that stage.